File: scripts/core_vectors.py
# SPDX-FileCopyrightText: 2026 Edward Kmett
# SPDX-License-Identifier: UPL-1.0 AND BSD-3-Clause

"""Exact bounded vector proof checks shared by Core audit sites."""
import json
from pathlib import Path

VECTOR_REP = {'kind': 'vector', 'primReps': ['VecRep 2 Int64ElemRep'], 'evaluated': True,
              'vector': {'lanes': 2, 'element': 'Int64ElemRep'}}
LANE_REP = {'kind': 'long', 'primReps': ['Int64Rep'], 'evaluated': True}
TUPLE_REP = {'kind': 'unknown', 'primReps': ['Int64Rep', 'Int64Rep'], 'evaluated': True,
             'aggregate': 'unboxed-tuple', 'components': [LANE_REP, LANE_REP]}
OPERATIONS = {
    'packInt64X2#': ([TUPLE_REP], VECTOR_REP),
    'unpackInt64X2#': ([VECTOR_REP], TUPLE_REP),
    'broadcastInt64X2#': ([LANE_REP], VECTOR_REP),
    'plusInt64X2#': ([VECTOR_REP, VECTOR_REP], VECTOR_REP),
    'minusInt64X2#': ([VECTOR_REP, VECTOR_REP], VECTOR_REP),
    'negateInt64X2#': ([VECTOR_REP], VECTOR_REP),
}
VECTOR32_REP = {'kind': 'vector', 'primReps': ['VecRep 4 Int32ElemRep'], 'evaluated': True,
                'vector': {'lanes': 4, 'element': 'Int32ElemRep'}}
LANE32_REP = {'kind': 'long', 'primReps': ['Int32Rep'], 'evaluated': True}
TUPLE32_REP = {'kind': 'unknown', 'primReps': ['Int32Rep'] * 4, 'evaluated': True,
               'aggregate': 'unboxed-tuple', 'components': [LANE32_REP] * 4}
OPERATIONS.update({
    'packInt32X4#': ([TUPLE32_REP], VECTOR32_REP),
    'unpackInt32X4#': ([VECTOR32_REP], TUPLE32_REP),
    'broadcastInt32X4#': ([LANE32_REP], VECTOR32_REP),
    'plusInt32X4#': ([VECTOR32_REP, VECTOR32_REP], VECTOR32_REP),
    'minusInt32X4#': ([VECTOR32_REP, VECTOR32_REP], VECTOR32_REP),
    'negateInt32X4#': ([VECTOR32_REP], VECTOR32_REP),
    'timesInt32X4#': ([VECTOR32_REP, VECTOR32_REP], VECTOR32_REP),
})
VECTOR16_REP = {'kind': 'vector', 'primReps': ['VecRep 8 Int16ElemRep'], 'evaluated': True,
                'vector': {'lanes': 8, 'element': 'Int16ElemRep'}}
LANE16_REP = {'kind': 'long', 'primReps': ['Int16Rep'], 'evaluated': True}
TUPLE16_REP = {'kind': 'unknown', 'primReps': ['Int16Rep'] * 8, 'evaluated': True,
               'aggregate': 'unboxed-tuple', 'components': [LANE16_REP] * 8}
OPERATIONS.update({
    'packInt16X8#': ([TUPLE16_REP], VECTOR16_REP),
    'unpackInt16X8#': ([VECTOR16_REP], TUPLE16_REP),
    'broadcastInt16X8#': ([LANE16_REP], VECTOR16_REP),
    'plusInt16X8#': ([VECTOR16_REP, VECTOR16_REP], VECTOR16_REP),
    'minusInt16X8#': ([VECTOR16_REP, VECTOR16_REP], VECTOR16_REP),
    'negateInt16X8#': ([VECTOR16_REP], VECTOR16_REP),
    'timesInt16X8#': ([VECTOR16_REP, VECTOR16_REP], VECTOR16_REP),
})
VECTOR8_REP = {'kind': 'vector', 'primReps': ['VecRep 16 Int8ElemRep'], 'evaluated': True,
               'vector': {'lanes': 16, 'element': 'Int8ElemRep'}}
LANE8_REP = {'kind': 'long', 'primReps': ['Int8Rep'], 'evaluated': True}
TUPLE8_REP = {'kind': 'unknown', 'primReps': ['Int8Rep'] * 16, 'evaluated': True,
              'aggregate': 'unboxed-tuple', 'components': [LANE8_REP] * 16}
OPERATIONS.update({
    'packInt8X16#': ([TUPLE8_REP], VECTOR8_REP),
    'unpackInt8X16#': ([VECTOR8_REP], TUPLE8_REP),
    'broadcastInt8X16#': ([LANE8_REP], VECTOR8_REP),
    'plusInt8X16#': ([VECTOR8_REP, VECTOR8_REP], VECTOR8_REP),
    'minusInt8X16#': ([VECTOR8_REP, VECTOR8_REP], VECTOR8_REP),
    'negateInt8X16#': ([VECTOR8_REP], VECTOR8_REP),
    'timesInt8X16#': ([VECTOR8_REP, VECTOR8_REP], VECTOR8_REP),
})
VECTOR_WORD8_REP = {'kind': 'vector', 'primReps': ['VecRep 16 Word8ElemRep'], 'evaluated': True,
                    'vector': {'lanes': 16, 'element': 'Word8ElemRep'}}
LANE_WORD8_REP = {'kind': 'long', 'primReps': ['Word8Rep'], 'evaluated': True}
TUPLE_WORD8_REP = {'kind': 'unknown', 'primReps': ['Word8Rep'] * 16, 'evaluated': True,
                  'aggregate': 'unboxed-tuple', 'components': [LANE_WORD8_REP] * 16}
OPERATIONS.update({
    'packWord8X16#': ([TUPLE_WORD8_REP], VECTOR_WORD8_REP),
    'unpackWord8X16#': ([VECTOR_WORD8_REP], TUPLE_WORD8_REP),
    'broadcastWord8X16#': ([LANE_WORD8_REP], VECTOR_WORD8_REP),
    'plusWord8X16#': ([VECTOR_WORD8_REP, VECTOR_WORD8_REP], VECTOR_WORD8_REP),
    'minusWord8X16#': ([VECTOR_WORD8_REP, VECTOR_WORD8_REP], VECTOR_WORD8_REP),
    'timesWord8X16#': ([VECTOR_WORD8_REP, VECTOR_WORD8_REP], VECTOR_WORD8_REP),
})
VECTOR_WORD16_REP = {'kind': 'vector', 'primReps': ['VecRep 8 Word16ElemRep'], 'evaluated': True,
                    'vector': {'lanes': 8, 'element': 'Word16ElemRep'}}
LANE_WORD16_REP = {'kind': 'long', 'primReps': ['Word16Rep'], 'evaluated': True}
TUPLE_WORD16_REP = {'kind': 'unknown', 'primReps': ['Word16Rep'] * 8, 'evaluated': True,
                  'aggregate': 'unboxed-tuple', 'components': [LANE_WORD16_REP] * 8}
OPERATIONS.update({
    'packWord16X8#': ([TUPLE_WORD16_REP], VECTOR_WORD16_REP),
    'unpackWord16X8#': ([VECTOR_WORD16_REP], TUPLE_WORD16_REP),
    'broadcastWord16X8#': ([LANE_WORD16_REP], VECTOR_WORD16_REP),
    'plusWord16X8#': ([VECTOR_WORD16_REP, VECTOR_WORD16_REP], VECTOR_WORD16_REP),
    'minusWord16X8#': ([VECTOR_WORD16_REP, VECTOR_WORD16_REP], VECTOR_WORD16_REP),
    'timesWord16X8#': ([VECTOR_WORD16_REP, VECTOR_WORD16_REP], VECTOR_WORD16_REP),
})
VECTOR_WORD32_REP = {'kind': 'vector', 'primReps': ['VecRep 4 Word32ElemRep'], 'evaluated': True,
                    'vector': {'lanes': 4, 'element': 'Word32ElemRep'}}
LANE_WORD32_REP = {'kind': 'long', 'primReps': ['Word32Rep'], 'evaluated': True}
TUPLE_WORD32_REP = {'kind': 'unknown', 'primReps': ['Word32Rep'] * 4, 'evaluated': True,
                  'aggregate': 'unboxed-tuple', 'components': [LANE_WORD32_REP] * 4}
OPERATIONS.update({
    'packWord32X4#': ([TUPLE_WORD32_REP], VECTOR_WORD32_REP),
    'unpackWord32X4#': ([VECTOR_WORD32_REP], TUPLE_WORD32_REP),
    'broadcastWord32X4#': ([LANE_WORD32_REP], VECTOR_WORD32_REP),
    'plusWord32X4#': ([VECTOR_WORD32_REP, VECTOR_WORD32_REP], VECTOR_WORD32_REP),
    'minusWord32X4#': ([VECTOR_WORD32_REP, VECTOR_WORD32_REP], VECTOR_WORD32_REP),
    'timesWord32X4#': ([VECTOR_WORD32_REP, VECTOR_WORD32_REP], VECTOR_WORD32_REP),
})
VECTOR_FLOAT_REP = {'kind': 'vector', 'primReps': ['VecRep 4 FloatElemRep'], 'evaluated': True,
                    'vector': {'lanes': 4, 'element': 'FloatElemRep'}}
LANE_FLOAT_REP = {'kind': 'float', 'primReps': ['FloatRep'], 'evaluated': True}
TUPLE_FLOAT_REP = {'kind': 'unknown', 'primReps': ['FloatRep'] * 4, 'evaluated': True,
                   'aggregate': 'unboxed-tuple', 'components': [LANE_FLOAT_REP] * 4}
OPERATIONS.update({
    'packFloatX4#': ([TUPLE_FLOAT_REP], VECTOR_FLOAT_REP),
    'unpackFloatX4#': ([VECTOR_FLOAT_REP], TUPLE_FLOAT_REP),
    'broadcastFloatX4#': ([LANE_FLOAT_REP], VECTOR_FLOAT_REP),
    'plusFloatX4#': ([VECTOR_FLOAT_REP, VECTOR_FLOAT_REP], VECTOR_FLOAT_REP),
    'minusFloatX4#': ([VECTOR_FLOAT_REP, VECTOR_FLOAT_REP], VECTOR_FLOAT_REP),
    'timesFloatX4#': ([VECTOR_FLOAT_REP, VECTOR_FLOAT_REP], VECTOR_FLOAT_REP),
})
VECTOR_DOUBLE_REP = {'kind': 'vector', 'primReps': ['VecRep 2 DoubleElemRep'], 'evaluated': True,
                    'vector': {'lanes': 2, 'element': 'DoubleElemRep'}}
LANE_DOUBLE_REP = {'kind': 'double', 'primReps': ['DoubleRep'], 'evaluated': True}
TUPLE_DOUBLE_REP = {'kind': 'unknown', 'primReps': ['DoubleRep'] * 2, 'evaluated': True,
                   'aggregate': 'unboxed-tuple', 'components': [LANE_DOUBLE_REP] * 2}
OPERATIONS.update({
    'packDoubleX2#': ([TUPLE_DOUBLE_REP], VECTOR_DOUBLE_REP),
    'unpackDoubleX2#': ([VECTOR_DOUBLE_REP], TUPLE_DOUBLE_REP),
    'broadcastDoubleX2#': ([LANE_DOUBLE_REP], VECTOR_DOUBLE_REP),
    'plusDoubleX2#': ([VECTOR_DOUBLE_REP, VECTOR_DOUBLE_REP], VECTOR_DOUBLE_REP),
    'minusDoubleX2#': ([VECTOR_DOUBLE_REP, VECTOR_DOUBLE_REP], VECTOR_DOUBLE_REP),
    'timesDoubleX2#': ([VECTOR_DOUBLE_REP, VECTOR_DOUBLE_REP], VECTOR_DOUBLE_REP),
})

# The same pinned declarative table drives concrete JVM source generation. This
# table describes exact signatures; core-capabilities still gates availability.
GENERATED_SHAPES = []
for family in json.loads((Path(__file__).with_name('simd-families.json')).read_text())['families']:
    shape = dict(lanes=family['lanes'], element=family['element'])
    vector = dict(kind='vector', evaluated=True, primReps=[f"VecRep {shape['lanes']} {shape['element']}"], vector=shape)
    lane = dict(kind={'FloatRep': 'float', 'DoubleRep': 'double'}.get(family['laneRep'], 'long'),
                evaluated=True, primReps=[family['laneRep']])
    packed = dict(kind='unknown', evaluated=True, aggregate='unboxed-tuple',
                  primReps=lane['primReps'] * family['lanes'], components=[lane] * family['lanes'])
    if family['newCarrier']:
        GENERATED_SHAPES.append(shape)
    for operation in family['operations']:
        arguments = [packed] if operation == 'pack' else [lane] if operation == 'broadcast' else [vector, lane, dict(kind='long', evaluated=True, primReps=['IntRep'])] if operation == 'insert' else [vector] * (2 if operation in ('plus', 'minus', 'times', 'divide') else 1)
        OPERATIONS[operation + family['name'] + '#'] = (arguments, packed if operation == 'unpack' else vector)

def is_vector(rep): return isinstance(rep, dict) and rep.get('kind') == 'vector'
def signature_matches(expected, actual):
    """A primop signature requires concrete carriers, including every tuple lane."""
    if not isinstance(actual, dict): return False
    if any(expected.get(key) != actual.get(key) for key in ('kind', 'primReps', 'aggregate', 'vector')): return False
    children, other = expected.get('components'), actual.get('components')
    if children is None: return other is None
    return isinstance(other, list) and len(children) == len(other) and all(signature_matches(a, b) for a, b in zip(children, other))
def proof_error(rep):
    registers = rep.get('primReps')
    has_vector = isinstance(registers, list) and any(isinstance(r, str) and r.startswith('VecRep ') for r in registers)
    if not is_vector(rep):
        return 'Vector representation lacks exact vector metadata' if 'vector' in rep or has_vector and 'aggregate' not in rep else None
    shape = rep.get('vector')
    if not isinstance(shape, dict) or type(shape.get('lanes')) is not int or not isinstance(shape.get('element'), str) or 'aggregate' in rep:
        return 'Invalid Core vector shape'
    if registers != [f"VecRep {shape['lanes']} {shape['element']}"]:
        return 'Vector shape disagrees with primitive representation'
    if shape not in (VECTOR_REP['vector'], VECTOR32_REP['vector'], VECTOR16_REP['vector'], VECTOR8_REP['vector'], VECTOR_WORD8_REP['vector'], VECTOR_WORD16_REP['vector'], VECTOR_WORD32_REP['vector'], VECTOR_FLOAT_REP['vector'], VECTOR_DOUBLE_REP['vector'], *GENERATED_SHAPES):
        return 'Unsupported Core vector representation'
    return None

File: src/main/kotlin/thc/runtime/CoreVectors.kt
// SPDX-FileCopyrightText: 2026 Edward Kmett
// SPDX-License-Identifier: UPL-1.0 AND BSD-3-Clause

package thc.runtime

import com.oracle.truffle.api.CompilerDirectives
import jdk.incubator.vector.*

/** Vector identity is one VecRep, independent of its physical runtime value. */
internal data class CoreVector(val lanes: Int, val element: String) {
    companion object {
        val INT64X2 = CoreVector(2, "Int64ElemRep")
        val INT32X4 = CoreVector(4, "Int32ElemRep")
        val INT16X8 = CoreVector(8, "Int16ElemRep")
        val INT8X16 = CoreVector(16, "Int8ElemRep")
        val WORD8X16 = CoreVector(16, "Word8ElemRep")
        val WORD16X8 = CoreVector(8, "Word16ElemRep")
        val WORD32X4 = CoreVector(4, "Word32ElemRep")
        val FLOATX4 = CoreVector(4, "FloatElemRep")
        val DOUBLEX2 = CoreVector(2, "DoubleElemRep")
        fun parse(raw: Any?, kind: CoreKind, reps: List<String>?, aggregate: Boolean): CoreVector? {
            if (kind != CoreKind.VECTOR) {
                // The exporter records a sole physical VecRep on a logical
                // tuple as well (for example (# State#, FloatX4# #)). Check that
                // redundant physical annotation without turning the tuple into
                // a scalar vector; its recursive fields are validated separately.
                if (aggregate && raw != null) {
                    parse(raw, CoreKind.VECTOR, reps, false)
                    return null
                }
                if (raw != null || !aggregate && reps?.any { it.startsWith("VecRep ") } == true)
                    throw RuntimeFault("Vector representation lacks exact vector metadata")
                return null
            }
            val record = raw as? Map<*, *> ?: throw RuntimeFault("Missing Core vector shape")
            val lanes = record["lanes"] as? Number ?: throw RuntimeFault("Invalid vector lane count")
            if (lanes !is Int && lanes !is Long) throw RuntimeFault("Invalid vector lane count")
            val element = record["element"] as? String ?: throw RuntimeFault("Invalid vector element kind")
            if (lanes.toDouble() != lanes.toInt().toDouble() || aggregate)
                throw RuntimeFault("Invalid Core vector shape")
            val vector = CoreVector(lanes.toInt(), element)
            if (reps != listOf("VecRep ${vector.lanes} $element")) throw RuntimeFault("Vector shape disagrees with primitive representation")
            if (vector != INT64X2 && vector != INT32X4 && vector != INT16X8 && vector != INT8X16 && vector != WORD8X16 && vector != WORD16X8 && vector != WORD32X4 && vector != FLOATX4 && vector != DOUBLEX2 && !GeneratedVectors.supports(vector)) throw UnsupportedCore("Unsupported Core vector representation: $vector")
            return vector
        }
    }
}

internal object CoreVectors {
    // Generic value boundaries check physical species; signedness stays in the exact VecRep.
    // Expose the species' exact class to the compiler without naming its inaccessible implementation.
    @JvmStatic fun requireByte(raw: Any?, species: VectorSpecies<Byte>): ByteVector {
        val value = raw as? ByteVector ?: fault("Expected ByteVector")
        if (value.species() != species) fault("Unexpected vector species")
        return CompilerDirectives.castExact(value, species.vectorType()) as ByteVector
    }
    @JvmStatic fun requireShort(raw: Any?, species: VectorSpecies<Short>): ShortVector {
        val value = raw as? ShortVector ?: fault("Expected ShortVector")
        if (value.species() != species) fault("Unexpected vector species")
        return CompilerDirectives.castExact(value, species.vectorType()) as ShortVector
    }
    @JvmStatic fun requireInt(raw: Any?, species: VectorSpecies<Int>): IntVector {
        val value = raw as? IntVector ?: fault("Expected IntVector")
        if (value.species() != species) fault("Unexpected vector species")
        return CompilerDirectives.castExact(value, species.vectorType()) as IntVector
    }
    @JvmStatic fun requireLong(raw: Any?, species: VectorSpecies<Long>): LongVector {
        val value = raw as? LongVector ?: fault("Expected LongVector")
        if (value.species() != species) fault("Unexpected vector species")
        return CompilerDirectives.castExact(value, species.vectorType()) as LongVector
    }
    @JvmStatic fun requireFloat(raw: Any?, species: VectorSpecies<Float>): FloatVector {
        val value = raw as? FloatVector ?: fault("Expected FloatVector")
        if (value.species() != species) fault("Unexpected vector species")
        return CompilerDirectives.castExact(value, species.vectorType()) as FloatVector
    }
    @JvmStatic fun requireDouble(raw: Any?, species: VectorSpecies<Double>): DoubleVector {
        val value = raw as? DoubleVector ?: fault("Expected DoubleVector")
        if (value.species() != species) fault("Unexpected vector species")
        return CompilerDirectives.castExact(value, species.vectorType()) as DoubleVector
    }
    @JvmStatic fun laneIndex(index: Long, lanes: Int): Int {
        if (index < 0L || index >= lanes.toLong()) fault("Invalid vector lane index")
        return index.toInt()
    }

    /** GHC's shuffle instruction requires a literal tuple of concatenated lane indices. */
    fun shuffleIndices(raw: List<Any?>, lanes: Int): IntArray {
        val constructor = raw.getOrNull(1) as? List<*>
        val fields = raw.getOrNull(2) as? List<*>
        if (raw.firstOrNull() != "app" || constructor?.firstOrNull() != "con" ||
            raw.getOrNull(5) != true || fields?.size != lanes)
            throw UnsupportedCore("Vector shuffle requires a literal index tuple")
        return IntArray(lanes) { lane ->
            val literal = fields[lane] as? List<*>
            val value = if (literal?.firstOrNull() == "lit" && literal.getOrNull(1) == "int")
                (literal.getOrNull(2) as? String)?.toLongOrNull() else null
            if (value == null || value < 0L || value >= 2L * lanes)
                throw UnsupportedCore("Vector shuffle indices must be literals in 0 until ${2 * lanes}")
            value.toInt()
        }
    }

    val proof = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 2 Int64ElemRep"), vector = CoreVector.INT64X2)
    private val lane = CoreRepresentation(CoreKind.LONG, true, true, listOf("Int64Rep"))
    val unpacked = CoreRepresentation(CoreKind.UNKNOWN, true, true, listOf("Int64Rep", "Int64Rep"), listOf(lane, lane))
    val proof32 = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 4 Int32ElemRep"), vector = CoreVector.INT32X4)
    private val lane32 = CoreRepresentation(CoreKind.LONG, true, true, listOf("Int32Rep"))
    val unpacked32 = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(4) { "Int32Rep" }, List(4) { lane32 })
    val proof16 = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 8 Int16ElemRep"), vector = CoreVector.INT16X8)
    private val lane16 = CoreRepresentation(CoreKind.LONG, true, true, listOf("Int16Rep"))
    val unpacked16 = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(8) { "Int16Rep" }, List(8) { lane16 })
    val operations16 = setOf("packInt16X8#", "unpackInt16X8#", "broadcastInt16X8#", "plusInt16X8#", "minusInt16X8#", "negateInt16X8#", "timesInt16X8#")
    val proof8 = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 16 Int8ElemRep"), vector = CoreVector.INT8X16)
    private val lane8 = CoreRepresentation(CoreKind.LONG, true, true, listOf("Int8Rep"))
    val unpacked8 = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(16) { "Int8Rep" }, List(16) { lane8 })
    val operations8 = setOf("packInt8X16#", "unpackInt8X16#", "broadcastInt8X16#", "plusInt8X16#", "minusInt8X16#", "negateInt8X16#", "timesInt8X16#")
    val proofWord8 = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 16 Word8ElemRep"), vector = CoreVector.WORD8X16)
    private val laneWord8 = CoreRepresentation(CoreKind.LONG, true, true, listOf("Word8Rep"))
    val unpackedWord8 = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(16) { "Word8Rep" }, List(16) { laneWord8 })
    val operationsWord8 = setOf("packWord8X16#", "unpackWord8X16#", "broadcastWord8X16#", "plusWord8X16#", "minusWord8X16#", "timesWord8X16#")
    val proofWord16 = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 8 Word16ElemRep"), vector = CoreVector.WORD16X8)
    private val laneWord16 = CoreRepresentation(CoreKind.LONG, true, true, listOf("Word16Rep"))
    val unpackedWord16 = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(8) { "Word16Rep" }, List(8) { laneWord16 })
    val operationsWord16 = setOf("packWord16X8#", "unpackWord16X8#", "broadcastWord16X8#", "plusWord16X8#", "minusWord16X8#", "timesWord16X8#")
    val proofWord32 = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 4 Word32ElemRep"), vector = CoreVector.WORD32X4)
    private val laneWord32 = CoreRepresentation(CoreKind.LONG, true, true, listOf("Word32Rep"))
    val unpackedWord32 = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(4) { "Word32Rep" }, List(4) { laneWord32 })
    val operationsWord32 = setOf("packWord32X4#", "unpackWord32X4#", "broadcastWord32X4#", "plusWord32X4#", "minusWord32X4#", "timesWord32X4#")
    val proofFloat = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 4 FloatElemRep"), vector = CoreVector.FLOATX4)
    private val laneFloat = CoreRepresentation(CoreKind.FLOAT, true, true, listOf("FloatRep"))
    val unpackedFloat = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(4) { "FloatRep" }, List(4) { laneFloat })
    val fusedFloat = listOf("fmaddFloatX4#", "fmsubFloatX4#", "fnmaddFloatX4#", "fnmsubFloatX4#")
    val fusedFloat8 = listOf("fmaddFloatX8#", "fmsubFloatX8#", "fnmaddFloatX8#", "fnmsubFloatX8#")
    val fusedFloat16 = listOf("fmaddFloatX16#", "fmsubFloatX16#", "fnmaddFloatX16#", "fnmsubFloatX16#")
    val operationsFloat = setOf("packFloatX4#", "unpackFloatX4#", "broadcastFloatX4#", "plusFloatX4#", "minusFloatX4#", "timesFloatX4#") + fusedFloat
    val proofDouble = CoreRepresentation(CoreKind.VECTOR, true, true, listOf("VecRep 2 DoubleElemRep"), vector = CoreVector.DOUBLEX2)
    private val laneDouble = CoreRepresentation(CoreKind.DOUBLE, true, true, listOf("DoubleRep"))
    val unpackedDouble = CoreRepresentation(CoreKind.UNKNOWN, true, true, List(2) { "DoubleRep" }, List(2) { laneDouble })
    val fusedDouble = listOf("fmaddDoubleX2#", "fmsubDoubleX2#", "fnmaddDoubleX2#", "fnmsubDoubleX2#")
    val fusedDouble4 = listOf("fmaddDoubleX4#", "fmsubDoubleX4#", "fnmaddDoubleX4#", "fnmsubDoubleX4#")
    val fusedDouble8 = listOf("fmaddDoubleX8#", "fmsubDoubleX8#", "fnmaddDoubleX8#", "fnmsubDoubleX8#")
    val operationsDouble = setOf("packDoubleX2#", "unpackDoubleX2#", "broadcastDoubleX2#", "plusDoubleX2#", "minusDoubleX2#", "timesDoubleX2#") + fusedDouble
    val operations32 = setOf("packInt32X4#", "unpackInt32X4#", "broadcastInt32X4#", "plusInt32X4#", "minusInt32X4#", "negateInt32X4#", "timesInt32X4#")
    val operations = setOf("packInt64X2#", "unpackInt64X2#", "broadcastInt64X2#", "plusInt64X2#", "minusInt64X2#", "negateInt64X2#") + operations32 + operations16 + operations8 + operationsWord8 + operationsWord16 + operationsWord32 + operationsFloat + operationsDouble + GeneratedVectors.operations + fusedFloat8 + fusedFloat16 + fusedDouble4 + fusedDouble8
    fun requireVariableProof(binding: CoreRepresentation?, occurrence: CoreRepresentation) {
        if (occurrence.isVector && binding?.vector != occurrence.vector)
            throw RuntimeFault("Vector occurrence lacks a matching lexical binder proof")
    }
    /** A local case must retain vector identity even when its result annotation is absent. */
    fun caseResult(alternatives: List<CoreRepresentation>): CoreRepresentation? {
        val vector = alternatives.firstOrNull { it.isVector } ?: return null
        if (alternatives.any { it.vector != vector.vector })
            throw RuntimeFault("Conflicting vector case result representations")
        return vector.copy(evaluated = alternatives.all { it.evaluated })
    }
    private fun exact(expected: CoreRepresentation, actual: CoreRepresentation): Boolean =
        actual.present && expected.kind == actual.kind && expected.vector == actual.vector &&
            TupleShape.compatible(expected, actual) && (expected.components == null ||
                expected.components.indices.all { exact(expected.components[it], actual.components!![it]) })
    fun validate(name: String, arguments: List<CoreRepresentation>, result: CoreRepresentation) {
        if (name in GeneratedVectors.operations) return GeneratedVectors.validate(name, arguments, result)
        val expected = when (name) {
            "packInt64X2#" -> listOf(unpacked)
            "unpackInt64X2#", "negateInt64X2#" -> listOf(proof)
            "broadcastInt64X2#" -> listOf(lane)
            "plusInt64X2#", "minusInt64X2#" -> listOf(proof, proof)
            "packInt32X4#" -> listOf(unpacked32)
            "unpackInt32X4#", "negateInt32X4#" -> listOf(proof32)
            "broadcastInt32X4#" -> listOf(lane32)
            "plusInt32X4#", "minusInt32X4#", "timesInt32X4#" -> listOf(proof32, proof32)
            "packInt16X8#" -> listOf(unpacked16)
            "unpackInt16X8#", "negateInt16X8#" -> listOf(proof16)
            "broadcastInt16X8#" -> listOf(lane16)
            "plusInt16X8#", "minusInt16X8#", "timesInt16X8#" -> listOf(proof16, proof16)
            "packInt8X16#" -> listOf(unpacked8)
            "unpackInt8X16#", "negateInt8X16#" -> listOf(proof8)
            "broadcastInt8X16#" -> listOf(lane8)
            "plusInt8X16#", "minusInt8X16#", "timesInt8X16#" -> listOf(proof8, proof8)
            "packWord8X16#" -> listOf(unpackedWord8)
            "unpackWord8X16#" -> listOf(proofWord8)
            "broadcastWord8X16#" -> listOf(laneWord8)
            "plusWord8X16#", "minusWord8X16#", "timesWord8X16#" -> listOf(proofWord8, proofWord8)
            "packWord16X8#" -> listOf(unpackedWord16)
            "unpackWord16X8#" -> listOf(proofWord16)
            "broadcastWord16X8#" -> listOf(laneWord16)
            "plusWord16X8#", "minusWord16X8#", "timesWord16X8#" -> listOf(proofWord16, proofWord16)
            "packWord32X4#" -> listOf(unpackedWord32)
            "unpackWord32X4#" -> listOf(proofWord32)
            "broadcastWord32X4#" -> listOf(laneWord32)
            "plusWord32X4#", "minusWord32X4#", "timesWord32X4#" -> listOf(proofWord32, proofWord32)
            "packDoubleX2#" -> listOf(unpackedDouble)
            "unpackDoubleX2#" -> listOf(proofDouble)
            "broadcastDoubleX2#" -> listOf(laneDouble)
            "plusDoubleX2#", "minusDoubleX2#", "timesDoubleX2#" -> listOf(proofDouble, proofDouble)
            "packFloatX4#" -> listOf(unpackedFloat)
            "unpackFloatX4#" -> listOf(proofFloat)
            "broadcastFloatX4#" -> listOf(laneFloat)
            "plusFloatX4#", "minusFloatX4#", "timesFloatX4#" -> listOf(proofFloat, proofFloat)
            in fusedFloat8 -> List(3) { GeneratedVectors.proofFloatX8 }
            in fusedFloat16 -> List(3) { GeneratedVectors.proofFloatX16 }
            in fusedDouble4 -> List(3) { GeneratedVectors.proofDoubleX4 }
            in fusedDouble8 -> List(3) { GeneratedVectors.proofDoubleX8 }
            in fusedFloat -> listOf(proofFloat, proofFloat, proofFloat)
            in fusedDouble -> listOf(proofDouble, proofDouble, proofDouble)
            else -> throw UnsupportedCore("Unsupported vector primitive $name")
        }
        val expectedResult = when (name) {
            "unpackInt64X2#" -> unpacked
            "unpackInt32X4#" -> unpacked32
            "unpackInt16X8#" -> unpacked16
            in operations16 -> proof16
            "unpackInt8X16#" -> unpacked8
            in operations8 -> proof8
            "unpackWord8X16#" -> unpackedWord8
            in operationsWord8 -> proofWord8
            "unpackWord16X8#" -> unpackedWord16
            in operationsWord16 -> proofWord16
            "unpackWord32X4#" -> unpackedWord32
            in operationsWord32 -> proofWord32
            "unpackDoubleX2#" -> unpackedDouble
            in operationsDouble -> proofDouble
            "unpackFloatX4#" -> unpackedFloat
            in fusedFloat8 -> GeneratedVectors.proofFloatX8
            in fusedFloat16 -> GeneratedVectors.proofFloatX16
            in fusedDouble4 -> GeneratedVectors.proofDoubleX4
            in fusedDouble8 -> GeneratedVectors.proofDoubleX8
            in operationsFloat -> proofFloat
            in operations32 -> proof32
            else -> proof
        }
        validateSignature(name, arguments, result, expected, expectedResult)
    }
    fun validateSignature(name: String, arguments: List<CoreRepresentation>, result: CoreRepresentation,
                          expected: List<CoreRepresentation>, expectedResult: CoreRepresentation) {
        if (arguments.size != expected.size) throw RuntimeFault("Vector primitive arity mismatch: $name")
        arguments.indices.forEach { i ->
            if (!exact(expected[i], arguments[i]))
                throw RuntimeFault("Vector primitive argument representation mismatch: $name argument $i")
        }
        if (!exact(expectedResult, result))
            throw RuntimeFault("Vector primitive result representation mismatch: $name")
    }
    fun validateFlags(flags: List<*>) {
        if (flags.any { it != false }) throw RuntimeFault("Vector primitive operands must be unlifted")
    }
    fun argumentProof(expression: List<Any?>): CoreRepresentation =
        if (expression.firstOrNull() == "lit" && expression.getOrNull(1) in listOf("int8", "word8", "int16", "word16", "int32", "word32"))
            CoreRepresentations.narrowLiteralProof(expression)
        else CoreRepresentations.expression(expression)
}
